### ISSUE TITLE
ability to download recorded videos

### ISSUE THREAD

**Issue by blademckain** [rroller/dahua#97]
**Is your feature request related to a problem? Please describe.**
At the moment there is no home assistant integration that can interact with the dahua camera to view the recorded videos.
It would certainly be innovative to have a quick and easy way to access recorded videos

**Describe the solution you'd like**
Using API
4.11.13 Download Media File between Times
you can download the videos

**Additional context**
a working example here
https://mlog.club/article/3576624

**Comment by rroller** [comment]
Thanks for the example. Do you know of an example Home Assistant integration that allows you to download files? I'm not sure I've seen anything like that before (or to show video clips). One problematic thing is that the native files that Dahua produce are .dav files. DAV is a modified mp4 format and it's not going to play without a conversion. I most likely won't be able to support this but will keep an eye out if I think it can be done easily.

**Comment by GaryOkie** [comment]
Other than my AD410 doorbell, none of my Dahua cameras have an SD-Card as they record only on the NVR.   Unfortunately, I discovered that there is no "NewFile" event raised when a motion event occurs and the video is recorded to the NVR.   This seems to only occur if an SD_Card is installed.

Should there be an additional event handler to subscribe to NVR "channel 0" events, assuming there is such a thing?  How else can NVR-specific events such as the locally-recorded filename be captured that aren't just passed through from the camera?

Maybe the filename can be inferred from the timestamp on the trigger event, but that's iffy.

So if the OP wants to be able to download video from the SD-Card and not the NVR, there is a way forward that should not require any changes to the Dahua integration.

**Comment by GaryOkie** [comment]
More discovery...  the DAV files recorded on the Dahua NVR are from the constant stream of each camera channel.   The recordings are written as individual hour-long DAV files. If you use the NVR Backup to USB drive facility, you will see these hour-long recordings which can be saved as DAV (or MP4 in the newer firmware).

Here is an example of the NVR filepath/name created:
`FilePath=/mnt/dvr/2021-10-04/4/dav/09/0/3/438066/09.00.02-10.00.02[R][0@0][0].dav

You can search the NVR for available files, but it's tricky since it takes 3 separate commands which require a session object id that is temporary...

`<nvr_ip>/cgi-bin/mediaFileFind.cgi?action=factory.create`
returns <object_id>   (valid only for a minute)

`<nvr_ip>/cgi-bin/mediaFileFind.cgi?action=findFile&object=<object_id>&condition.Channel=5&condition.StartTime=2021-10-4%2010:00:00%20&condition.EndTime=2021-10-4%2012:00:00`

`<nvr_ip>/cgi-bin/mediaFileFind.cgi?action=findNextFile&object=<object_id>`

```
items[0].VideoStream=Main
items[0].Channel=4
items[0].Type=dav
items[0].StartTime=2021-10-04 09:00:02
items[0].EndTime=2021-10-04 10:00:02
items[0].Disk=3
items[0].Partition=3
items[0].Cluster=438066
items[0].FilePath=/mnt/dvr/2021-10-04/4/dav/09/0/3/438066/09.00.02-10.00.02[R][0@0][0].dav
items[0].Length=953417728
items[0].Flags[0]=Timing
items[0].CutLength=530432
```

I was not successful using the http RPC_Loadfile command to directly download this file.  It immediately fails with a "file incomplete" error.  Not that I wanted a huge hour long MP4 anyway.

Bottom line - a facility to download recorded videos from the NVR would be very complicated in my view.   On the other hand, accessing them from the camera's SD-Card is much simpler given the specific filename obtained from the "NewFile" event.

**Comment by blademckain** [comment]
Hi guys,
I report the configuration I used to solve this problem (if it were integrated in this component it would be fantastic, but I know it's complicated)

Prerequisite: u have to enable ffmpeg in your configuration.yaml

1. Automation for video download
```
input_text:
  last_recorded_video:
    name: Last Recorded Video
    max: 255
  last_snapshot_camera:
    name: Last SnapShot Camera
    max: 255

shell_command:
  download_recorded_dav: "curl --digest --user {{tvcc_user}}:{{tvcc_pass}} -o /config/www/tmp/last_recorded.dav -g {{tvcc_url}}"
  download_snapshot_camera: "curl --digest --user {{tvcc_user}}:{{tvcc_pass}} -o /config/www/tmp/last_snapshot.jpg -g {{tvcc_url}}"
  convert_recorded_dav: "bash /config/www/tmp/convertdav.sh"

    - alias: "Rilevazione video registrato - Telecamere"
      mode: queued
      variables:
        tvcc_user: !secret tvcc_user
        tvcc_pass: !secret tvcc_pass
      trigger:
         - platform: event
           event_type: dahua_event_received
           event_data:
              Code: NewFile
              action: Pulse
      condition:
        # Only trigger the alarm if it's armed
        condition: or
        conditions:
          - condition: state
            entity_id: alarm_control_panel.alarm
            state: armed_home
          - condition: state
            entity_id: alarm_control_panel.alarm
            state: armed_night
          - condition: state
            entity_id: alarm_control_panel.alarm
            state: armed_away
          - condition: state
            entity_id: alarm_control_panel.alarm
            state: triggered
      action:
        - choose:
            - conditions: "{{ '.dav' in trigger.event.data.data.File }}"
              sequence: 
                - service: input_text.set_value
                  data_template:
                    entity_id: input_text.last_recorded_video
                    value: >
                      {% if trigger.event.data.name == 'TVCC2' %}
                      {% set tvcc_ip = '192.168.x.44' %}
                      {% elif trigger.event.data.name == 'TVCC4' %}
                      {% set tvcc_ip = '192.168.x.42' %}
                      {% elif trigger.event.data.name == 'TVCC5' %}
                      {% set tvcc_ip = '192.168.x.41' %}
                      {% endif %}
                      {{ 'http://'~tvcc_ip~'/cgi-bin/RPC_Loadfile/'~trigger.event.data.data.File }}
                - service: shell_command.download_recorded_dav
                  data_template:
                    tvcc_user: "{{tvcc_user}}"
                    tvcc_pass: "{{tvcc_pass}}"
                    tvcc_url: "{{ states('input_text.last_recorded_video') }}"
                - service: shell_command.convert_recorded_dav
            - conditions: "{{ '.jpg' in trigger.event.data.data.File }}"
              sequence:
                - service: input_text.set_value
                  data_template:
                    entity_id: input_text.last_snapshot_camera
                    value: >
                      {% if trigger.event.data.name == 'TVCC2' %}
                      {% set tvcc_ip = '192.168.x.44' %}
                      {% elif trigger.event.data.name == 'TVCC4' %}
                      {% set tvcc_ip = '192.168.x.42' %}
                      {% elif trigger.event.data.name == 'TVCC5' %}
                      {% set tvcc_ip = '192.168.x.41' %}
                      {% endif %}
                      {{ 'http://'~tvcc_ip~'/cgi-bin/RPC_Loadfile/'~trigger.event.data.data.File }}
                - service: shell_command.download_snapshot_camera
                  data_template:
                    tvcc_user: "{{tvcc_user}}"
                    tvcc_pass: "{{tvcc_pass}}"
                    tvcc_url: "{{ states('input_text.last_snapshot_camera') }}"
          default: []
```

2. Bash script for convert video
```
#!/bin/bash
ffmpeg -y -i /config/www/tmp/last_recorded.dav -c copy -max_muxing_queue_size 9999 /config/www/tmp/last_recorded.mp4 2> /config/www/tmp/last_recorded.log
```

**Comment by rroller** [comment]
@blademckain  that's really cool. And yes, it's pretty complicated to build into the integration. Will keep this open for now and I'll think about it some more.

**Comment by tunip** [comment]
Have you already been able to think about it a little :-)?

**Comment by junalmeida** [comment]
@rroller Not sure if this helps but the native Nest integration can show recent video recordings on media source page of home assistant. 

![image](https://user-images.githubusercontent.com/991675/176595160-98143a00-f5f8-4e67-9034-25986690bcfa.png)

Xbox integration does that as well for my recorded game clips.

**Comment by augustodinizl** [comment]
> @rroller Not sure if this helps but the native Nest integration can show recent video recordings on media source page of home assistant. 
> 
> ![image](https://user-images.githubusercontent.com/991675/176595160-98143a00-f5f8-4e67-9034-25986690bcfa.png)
> 
> Xbox integration does that as well for my recorded game clips.

Tapo integration algo does that: https://github.com/JurajNyiri/HomeAssistant-Tapo-Control

**Comment by augustodinizl** [comment]
> Thanks for the example. Do you know of an example Home Assistant integration that allows you to download files? I'm not sure I've seen anything like that before (or to show video clips). One problematic thing is that the native files that Dahua produce are .dav files. DAV is a modified mp4 format and it's not going to play without a conversion. I most likely won't be able to support this but will keep an eye out if I think it can be done easily.

Man, i've already used that one (https://github.com/arthar360/TaniDVR ) with a rtsp mini server to stream the vídeo from a old dahua/intelbras DVR into home assistant, maybe it helps

**Comment by thedeadliestcatch** [comment]
This would be pretty useful to retrieve the associated file with the last event when possible, as camera snapshots will inevitably lag.

**Comment by dehooge** [comment]
Hey @blademckain is your solution still working? Trying to copy your automation but I can get it working. Probably something easy I'm missing but I'm an idiot at these things.
I get this error in logs, what did I miss to set up?

Template variable error: 'dict object' has no attribute 'event' when rendering '{{ '.dav' in trigger.event.data.data.File }}'
Template variable error: 'dict object' has no attribute 'event' when rendering '{{ '.jpg' in trigger.event.data.data.File }}'

Thanks

**Comment by augustodinizl** [comment]
> Thanks for the example. Do you know of an example Home Assistant integration that allows you to download files? I'm not sure I've seen anything like that before (or to show video clips). One problematic thing is that the native files that Dahua produce are .dav files. DAV is a modified mp4 format and it's not going to play without a conversion. I most likely won't be able to support this but will keep an eye out if I think it can be done easily.

in the https://github.com/arthar360/TaniDVR repo, are a plugin that converts dhav to mkv video, maybe it helps even better

**Comment by LuisPeregrina** [comment]
It would be nice to get the events in a list with snapshots and clips, like frigate and their lovelace card.

**Comment by jekkos** [comment]
It's not trivial to do this due to the way the recordings are stored. It's basically one file per hour that it saves and the events are then viewed in their software by opening this file at a specific position. Also the video format is a bit special but luckily it can be sliced using ffmpeg into smaller clips without having to reencode them.

I setup the NVR to copy the hourly recordings to my HA server. I have built a python script that I ran using systemd to slice these recordings using the event data in HA. The slices are then stored in a location where HA can access them through the media browser. So this way, the events have their corresponding clips available

**Comment by aman-agrawal** [comment]
Dahua API to download recorded videos from cctv/nvr given start dateTime and end dateTime in ubuntu ?
Why Dahua dont provide access to download videos via NVR webpage in ubuntu, can someone pls help.

**Comment by zios0007** [comment]
An honest status, because this issue keeps attracting people who then wait on it.

Nobody has built it, and it is larger than it looks. @jekkos put the reason above and it is the crux: the recorder stores a file per hour and events index into it at an offset, so "download the clip for this event" is not one API call. `mediaFileFind.cgi` lists files and `loadfile.cgi` fetches them, but turning that into something usable means a media source, clip extraction at an offset, and a way to browse events.

That is a feature on its own scale rather than something that lands alongside entity work.

Anyone who wants event clips with a card today is better served by Frigate, which does that job and runs alongside this integration rather than instead of it.

Leaving it open as a genuine request, but nobody should hold their breath.
